src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/RazorProjectChangePublisher.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
+{
+    internal abstract class RazorProjectChangePublisher : ProjectSnapshotChangeTrigger
```

Comment:
Out of curiosity why have this an abstract class with logic vs. all the logic sitting in the impl class?

Reply:
I felt that the logic which is kept on the abstract class is foundational enough to what it does that it wouldn't change depending on the implementation, and might therefore better serve as a framework for any future implementations to follow.

Reply:
> I felt that the logic which is kept on the abstract class is foundational enough to what it does that it wouldn't change depending on the implementation, and might therefore better serve as a framework for any future implementations to follow.

Ah, I wouldn't do that quite yet. My bet is that in the future we'll actually just use the same impl type in both.